minetest/minetest
# Add support for attached facedir/4dir nodes

## The situation

Currently, `attached_node` is useless if the node is a facedir or 4dir node. Your only option is to use wallmounted, but this has limitations. If facedir and 4dir nodes could be attached, that would allow game authors more flexibility.

## PR

This PR extends functionality of the `attached_node` group. Three new group ratings are added, for a new total of 4. This is how they work:

* `1`: Attaches to wallmounted direction if wallmounted, otherwise attaches to floor (current behavior, no change)
* `2`: Attaches to facedir/4dir direction if facedir/4dir (the "attaching side" is the fifth tile). Useless if the node is not facedir/4dir
* `3`: Always attaches to floor (even wallmounted. I admit this is quite esoteric, but I think it is better design to have a "neutral" option that completely ignores all other node fields to prevent annoying side effects)
* `4`: Always attaches to ceiling (because why not?)

## To do

It works so far for me, and needs review.

One thing I am not so sure about is placement prediction of attached facedir/4dir, please test this.

## How to test

Pick up the attached node variants from DevTest and check them out. Make sure the nodes stay attached when they should stay attached, and that they drop when they should not stay attached.

Because due to how facedir works, the facedir attached node can only be placed at the side of nodes and not floor or ceiling, because facedir nodes rotate only horizontally (param2=0..3) on placement. To test the other rotations, use the param2 tool.

## See also
#11431
